File: Eclipse/Java_Jna/src/main/java/com/wearesoft/Java_Jna/EntryPoint.java
package com.wearesoft.Java_Jna;

public class EntryPoint {

	public static void main(String[] args) {
		try {
			System.out.println("test func [add]");
			int r1 = Java_JnaPureCLibrary.instance.add(1, 2);
			System.out.println("result : " + r1);

			System.out.println("test func [get_order]");
			Java_JnaPureCLibrary.order.ByValue o1 = Java_JnaPureCLibrary.instance
					.get_order(3);
			Java_JnaPureCLibrary.order_item[] oi1 = (Java_JnaPureCLibrary.order_item[]) o1.order_items
					.toArray(o1.order_items_count);
			System.out.println("order_items_count : " + o1.order_items_count);
			for (int i = 0; i < o1.order_items_count; i++) {
				System.out.println("order_item[" + i + "]");
				System.out.println("order_count : " + oi1[i].order_count);
				System.out.println("name : " + oi1[i].order_item.name);
				System.out.println("weight : " + oi1[i].order_item.weight);
			}

			System.out.println("test func [get_order_return_point]");
			Java_JnaPureCLibrary.order.ByValue o2 = Java_JnaPureCLibrary.instance
					.get_order(2);
			Java_JnaPureCLibrary.order_item[] oi2 = (Java_JnaPureCLibrary.order_item[]) o2.order_items
					.toArray(o2.order_items_count);
			System.out.println("order_items_count : " + o2.order_items_count);
			for (int i = 0; i < o2.order_items_count; i++) {
				System.out.println("order_item[" + i + "]");
				System.out.println("order_count : " + oi2[i].order_count);
				System.out.println("name : " + oi2[i].order_item.name);
				System.out.println("weight : " + oi2[i].order_item.weight);
			}

			System.out.println("test func [change_order]");
			Java_JnaPureCLibrary.order.ByReference o3 = new Java_JnaPureCLibrary.order.ByReference();
			o3.order_items_count = 2;
			o3.order_items = new Java_JnaPureCLibrary.order_item.ByReference();
			Java_JnaPureCLibrary.order_item[] oi3 = (Java_JnaPureCLibrary.order_item[]) o3.order_items
					.toArray(o3.order_items_count);
			for (int i = 0; i < o3.order_items_count; i++) {
				oi3[i].writeField("order_count", i + 1);
				oi3[i].order_item.writeField("name", "order" + (i + 1));
				oi3[i].order_item.writeField("weight", (float) (i + 0.1));
			}
		
			System.out.println("order_items_count : " + o3.order_items_count);
			for (int i = 0; i < o3.order_items_count; i++) {
				System.out.println("order_item[" + i + "]");
				System.out.println("order_count : " + oi3[i].order_count);
				System.out.println("name : " + oi3[i].order_item.name);
				System.out.println("weight : " + oi3[i].order_item.weight);
			}

			Java_JnaPureCLibrary.order.ByValue o4 = Java_JnaPureCLibrary.instance
					.change_order(o3, (float) 0.9);
			System.out.println("result==========");
			Java_JnaPureCLibrary.order_item[] oi4 = (Java_JnaPureCLibrary.order_item[]) o4.order_items
					.toArray(o4.order_items_count);
			System.out.println("order_items_count : " + o4.order_items_count);
			for (int i = 0; i < o4.order_items_count; i++) {
				System.out.println("order_item[" + i + "]");
				System.out.println("order_count : " + oi4[i].order_count);
				System.out.println("name : " + oi4[i].order_item.name);
				System.out.println("weight : " + oi4[i].order_item.weight);
			}

		} catch (Exception e) {
			System.out.println(e.toString() + e.getMessage());
			return;
		}

		return;
	}
}

File: Eclipse/Java_JnaPureC/entry_point.c
/*
 * entry_point.c
 *
 *  Created on: 2014/7/13
 *      Author: administrator
 */

#include <stdio.h>
#include <func.h>

int main(void) {
	int r1 = add(1, 2);
	printf("test func [add]\n");
	printf("return : %d\n", r1);

	order o1;
	int order_LENGTH = 3;
	o1 = get_order(order_LENGTH);
	printf("test func [get_order]\n");
	printf("order_items_count : %d\n", o1.order_items_count);
	int i;
	for (i = 0; i < order_LENGTH; i++) {
		printf("order_item [%d]\n", i);
		printf("order_count : %d\n", o1.order_items[i].order_count);
		printf("name : %s\n", o1.order_items[i].order_item.name);
		printf("weight : %.1f\n", o1.order_items[i].order_item.weight);
	}

	order *o2 = get_order_return_point(order_LENGTH);
	printf("test func [get_order_return_point]\n");
	printf("order_items_count : %d\n", o2->order_items_count);
	for (i = 0; i < order_LENGTH; i++) {
		printf("order_item [%d]\n", i);
		printf("order_count : %d\n", o2->order_items[i].order_count);
		printf("name : %s\n", o2->order_items[i].order_item.name);
		printf("weight : %.1f\n", o2->order_items[i].order_item.weight);
	}

	order o3 = change_order(&o1, (float)0.9);
	printf("test func [change_order]\n");
	printf("order_items_count : %d\n", o3.order_items_count);
	for (i = 0; i < order_LENGTH; i++) {
		printf("order_item [%d]\n", i);
		printf("order_count : %d\n", o3.order_items[i].order_count);
		printf("name : %s\n", o3.order_items[i].order_item.name);
		printf("weight : %.1f\n", o3.order_items[i].order_item.weight);
	}

	return 0;
}
